pkg/controller/clone-controller.go

```diff
@@ -10,82 +10,165 @@ import (
 	}
 
 	// Bind the pvc SharedIndexInformer to the pvc queue
 	c.pvcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(obj)
-			if err == nil {
-				c.pvcQueue.AddRateLimited(key)
-			}
-		},
-		// this is triggered by an update or it will also be
-		// be triggered periodically even if no changes were made.
+		AddFunc: c.enqueuePVC,
 		UpdateFunc: func(old, new interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(new)
-			if err == nil {
-				c.pvcQueue.AddRateLimited(key)
-			}
+			c.enqueuePVC(new)
 		},
 	})
 
 	// Bind the pod SharedIndexInformer to the pod queue
 	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(obj)
-			if err == nil {
-				c.podQueue.AddRateLimited(key)
-			}
-		},
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(newObj)
-			if err == nil {
-				c.podQueue.AddRateLimited(key)
+		AddFunc: c.handlePodAdd,
+		UpdateFunc: func(old, new interface{}) {
+			newDepl := new.(*v1.Pod)
+			oldDepl := old.(*v1.Pod)
+			if newDepl.ResourceVersion == oldDepl.ResourceVersion {
+				// Periodic resync will send update events for all known PVCs.
+				// Two different versions of the same PVCs will always have different RVs.
+				return
 			}
+			c.handlePodUpdate(new)
 		},
+		DeleteFunc: c.handlePodDelete,
 	})
 
 	return c
 }
 
+func (c *CloneController) handlePodAdd(obj interface{}) {
+	c.handlePodObject(obj, "add")
+}
+func (c *CloneController) handlePodUpdate(obj interface{}) {
+	c.handlePodObject(obj, "update")
+}
+func (c *CloneController) handlePodDelete(obj interface{}) {
+	c.handlePodObject(obj, "delete")
+}
+
+func (c *CloneController) expectPodCreate(pvcKey string) {
+	c.podExpectations.ExpectCreations(pvcKey, 1)
+}
+func (c *CloneController) observePodCreate(pvcKey string) {
+	c.podExpectations.CreationObserved(pvcKey)
+}
+
+func (c *CloneController) handlePodObject(obj interface{}, verb string) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object, invalid type"))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+		glog.V(Vdebug).Infof("Recovered deleted object '%s' from tombstone", object.GetName())
+	}
+	glog.V(Vdebug).Infof("Processing object: %s", object.GetName())
+	if ownerRef := metav1.GetControllerOf(object); ownerRef != nil {
+		_, createdByUs := object.GetAnnotations()[AnnCloningCreatedBy]
+		_, isTargetPod := object.GetAnnotations()[AnnIsTargetPod]
+
+		if ownerRef.Kind != "PersistentVolumeClaim" {
+			return
+		} else if !createdByUs  || !isTargetPod {
+			return
```

Comment:
ditto.

Reply:
another optimization. I think i was just trying to add some easy checks before doing the pvc lookup in the cache.